[WebGPU Swift] CommandEncoder does not support GPUTexture instances in beginRenderPass

mwyrzykowski opened the pull request:
#### ab97aabb7139f6458361a641ff97e08637975ce6
<pre>
[WebGPU Swift] CommandEncoder does not support GPUTexture instances in beginRenderPass
<a href="https://bugs.webkit.org/show_bug.cgi?id=299456">https://bugs.webkit.org/show_bug.cgi?id=299456</a>
<a href="https://rdar.apple.com/161261339">rdar://161261339</a>

Reviewed by Cameron McCormack.

Factor out Texture variant into its own header file so Swift
and Objective-C implementations can share the same logic.

* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::CommandEncoder::beginRenderPass):
(WebGPU::TextureOrTextureView::TextureOrTextureView): Deleted.
(WebGPU::TextureOrTextureView::setCommandEncoder): Deleted.
(WebGPU::isRenderableTextureView): Deleted.
* Source/WebGPU/WebGPU/CommandEncoder.swift:
(WebGPU.beginRenderPass(_:)):
(WebGPU.isRenderableTextureView(_:)): Deleted.
* Source/WebGPU/WebGPU/Internal/WebGPUSwiftInternal.h:
(WebGPU_Internal::isValidToUseWith):
* Source/WebGPU/WebGPU/TextureOrTextureView.h: Added.
(WebGPU::TextureOrTextureView::TextureOrTextureView):
(WebGPU::TextureOrTextureView::setCommandEncoder):
(WebGPU::TextureOrTextureView::rasterizationMapForSlice):
(WebGPU::isRenderableTextureView):

Canonical link: <a href="https://commits.webkit.org/300997@main">https://commits.webkit.org/300997@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b2bc0ccdeec3ceadfd1141c2683c68935a8a515e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/124585 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/44257 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/34991 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/131424 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/76533 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/6f3b29a6-9838-4b7e-aaf6-d3d816408ec2) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/44961 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52828 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/94777 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/62851 "Passed tests") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/127539 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35843 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/111415 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75349 "Passed tests") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/84d548d4-4891-42e0-a6f1-5dc24e470513) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/34781 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/74903 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/105598 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29801 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/134092 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51437 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/39258 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/103255 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/51845 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107637 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/103034 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26238 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48396 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26659 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/48371 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/51308 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/57104 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/50711 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/54066 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52397 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

heycam commented on Source/WebGPU/WebGPU/CommandEncoder.mm:
```diff
@@ -738,6 +679,8 @@ static bool isRenderableTextureView(const auto& texture)
                 texture.setPreviouslyCleared();
             if (attachment.resolveTarget)
                 protectedFromAPI(attachment.resolveTarget)->setPreviouslyCleared();
+            if (attachment.resolveTexture)
+                protectedFromAPI(attachment.resolveTexture)->setPreviouslyCleared();
```
Is this a drive by fix for the C++ implementation? What's the impact of not doing this / should there be a test for this?

mwyrzykowski replied:
Yes it is :) Noticed when making the corresponding Swift change - https://github.com/WebKit/WebKit/pull/51686/files#r2400184367

The impact of not doing is previously written to texture contents would get zeroed during a partial update. 

The test should be added to https://gpuweb.github.io/cts/standalone/?q=webgpu:api,operation,resource_init,texture_zero:* but since we pass that one at 100% the CTS seems to be missing a case there

I filed https://bugs.webkit.org/show_bug.cgi?id=300053 to add that

heycam commented on Source/WebGPU/WebGPU/CommandEncoder.swift:
```diff
@@ -1121,7 +1135,10 @@ extension WebGPU.CommandEncoder {
         if let attachment = wgpuGetRenderPassDescriptorDepthSencilAttachment(descriptorSpan)?[0] {
```
Existing nit: this function has a typo in its name.

mwyrzykowski replied:
I will fix the naming in this PR since its only used in a limited number of places

mwyrzykowski commented on Source/WebGPU/WebGPU/CommandEncoder.swift:
```diff
@@ -1031,24 +1035,30 @@ extension WebGPU.CommandEncoder {
                         WebGPU.fromAPI(attachment.resolveTarget).setPreviouslyCleared(0, 0)
                     }
+                    if attachment.resolveTexture != nil {
+                        WebGPU.fromAPI(attachment.resolveTexture).setPreviouslyCleared()
```
corresponding Swift change

mwyrzykowski committed:
[WebGPU Swift] CommandEncoder does not support GPUTexture instances in beginRenderPass
https://bugs.webkit.org/show_bug.cgi?id=299456
rdar://161261339

Reviewed by Cameron McCormack.

Factor out Texture variant into its own header file so Swift
and Objective-C implementations can share the same logic.

* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::CommandEncoder::beginRenderPass):
(WebGPU::TextureOrTextureView::TextureOrTextureView): Deleted.
(WebGPU::TextureOrTextureView::setCommandEncoder): Deleted.
(WebGPU::isRenderableTextureView): Deleted.
* Source/WebGPU/WebGPU/CommandEncoder.swift:
(WebGPU.beginRenderPass(_:)):
(WebGPU.isRenderableTextureView(_:)): Deleted.
* Source/WebGPU/WebGPU/Internal/WebGPUSwiftInternal.h:
(WebGPU_Internal::isValidToUseWith):
* Source/WebGPU/WebGPU/TextureOrTextureView.h: Added.
(WebGPU::TextureOrTextureView::TextureOrTextureView):
(WebGPU::TextureOrTextureView::setCommandEncoder):
(WebGPU::TextureOrTextureView::rasterizationMapForSlice):
(WebGPU::isRenderableTextureView):

Canonical link: https://commits.webkit.org/300997@main
